src/machine-c-api.h

```diff
@@ -82,1894 +83,544 @@ typedef enum { // NOLINT(modernize-use-using)
+
+// -------------------------------------
+// Rolling back
+
+/// \brief Saves a snapshot of the machine.
+/// \param m Pointer to a valid machine instance.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_snapshot(cm_machine *m);
+
+/// \brief Performs commit of the machine, discarding last snapshot.
+/// \param m Pointer to a valid machine instance.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_commit(cm_machine *m);
+
+/// \brief Performs rollback of the machine, restoring last snapshot.
+/// \param m Pointer to a valid machine instance.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_rollback(cm_machine *m);
+
+// -------------------------------------
+// Running and stepping
 
 /// \brief Runs the machine until mcycle reaches mcycle_end or the machine halts.
-/// \param m Pointer to valid machine instance
-/// \param mcycle_end End cycle value
-/// \param break_reason Receives reason for machine run interruption when not NULL
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-CM_API int cm_machine_run(cm_machine *m, uint64_t mcycle_end, CM_BREAK_REASON *break_reason_result, char **err_msg);
+/// \param m Pointer to a valid machine instance.
+/// \param mcycle_end End cycle value.
+/// \param break_reason Receives reason for machine run interruption when not NULL.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_run(cm_machine *m, uint64_t mcycle_end, CM_BREAK_REASON *break_reason);
+
+/// \brief Runs the machine in the microarchitecture until the mcycle advances by 1
+/// or uarch_cycle counter reaches uarch_cycle_end.
+/// \param m Pointer to a valid machine instance.
+/// \param uarch_cycle_end End cycle value
+/// \param break_reason Receives status of machine run_uarch when not NULL.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_run_uarch(cm_machine *m, uint64_t uarch_cycle_end, CM_UARCH_BREAK_REASON *break_reason);
 
 /// \brief Runs the machine for one micro cycle logging all accesses to the state.
-/// \param m Pointer to valid machine instance
-/// \param log_type Type of access log to generate.
+/// \param m Pointer to a valid machine instance.
+/// \param log_type Type of access log to generate, must be bitwise OR of CM_ACCESS_LOG_TYPE enum.
+/// \param one_based Use 1-based indices when reporting errors.
+/// \param access_log Receives the state access log as a JSON string,
+/// remains valid until this same function is called again for the same machine.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_log_uarch_step(cm_machine *m, int log_type, bool one_based, const char **access_log);
```

Comment:
Should be `cm_log_step_uarch`, right? Because if there was a function to step the uarch without returning a log, it would have been `cm_step_uarch` instead `cm_uarch_step`. Other functions would change accordingly.